src/server/pfs/s3/master_test.go

```diff
@@ -416,6 +417,76 @@ func masterListObjectsRecursive(t *testing.T, pachClient *client.APIClient, mini
 	checkListObjects(t, ch, &startTime, &endTime, expectedFiles, []string{})
 }
 
+func masterListSystemRepoBuckets(t *testing.T, pachClient *client.APIClient, minioClient *minio.Client) {
+	repo := tu.UniqueString("listsystemrepo")
+	require.NoError(t, pachClient.CreateRepo(repo))
+	specRepo := client.NewSystemRepo(repo, pfs.SpecRepoType)
+	_, err := pachClient.PfsAPIClient.CreateRepo(pachClient.Ctx(), &pfs.CreateRepoRequest{Repo: specRepo})
+	require.NoError(t, err)
+	metaRepo := client.NewSystemRepo(repo, pfs.MetaRepoType)
+	_, err = pachClient.PfsAPIClient.CreateRepo(pachClient.Ctx(), &pfs.CreateRepoRequest{Repo: metaRepo})
+	require.NoError(t, err)
+
+	// create a master branch on each
+	require.NoError(t, pachClient.CreateBranch(repo, "master", "", "", nil))
+	_, err = pachClient.PfsAPIClient.CreateBranch(pachClient.Ctx(), &pfs.CreateBranchRequest{Branch: specRepo.NewBranch("master")})
+	require.NoError(t, err)
+	_, err = pachClient.PfsAPIClient.CreateBranch(pachClient.Ctx(), &pfs.CreateBranchRequest{Branch: metaRepo.NewBranch("master")})
+	require.NoError(t, err)
+
+	buckets, err := minioClient.ListBuckets()
+	require.NoError(t, err)
+
+	var hasMaster, hasMeta bool
+	for _, bucket := range buckets {
+		if bucket.Name == fmt.Sprintf("master.%s", repo) {
+			hasMaster = true
+		} else if bucket.Name == fmt.Sprintf("master.%s.%s", pfs.MetaRepoType, repo) {
+			hasMeta = true
+		} else {
+			require.NotEqual(t, bucket.Name, fmt.Sprintf("master.%s.%s", pfs.SpecRepoType, repo))
+		}
+	}
+
+	require.True(t, hasMaster)
+	require.True(t, hasMeta)
+}
+
+func masterResolveSystemRepoBucket(t *testing.T, pachClient *client.APIClient, minioClient *minio.Client) {
+	repo := tu.UniqueString("testsystemrepo")
+	require.NoError(t, pachClient.CreateRepo(repo))
+	specRepo := client.NewSystemRepo(repo, pfs.SpecRepoType)
+	// create a branch named "spec" in the repo
+	branch := pfs.SpecRepoType
+	require.NoError(t, pachClient.CreateBranch(repo, branch, "", "", nil))
+
+	// as well as a branch named "master" on an associated repo of type "spec"
+	_, err := pachClient.PfsAPIClient.CreateRepo(pachClient.Ctx(), &pfs.CreateRepoRequest{Repo: specRepo})
```

Comment:
very minor suggestion: move `specRepo` definition down to where `CreateRepo` is called:
```suggestion
	// create a branch named "spec" in the repo
	branch := pfs.SpecRepoType
	require.NoError(t, pachClient.CreateBranch(repo, branch, "", "", nil))

	// as well as a branch named "master" on an associated repo of type "spec"
	specRepo := client.NewSystemRepo(repo, pfs.SpecRepoType)
	_, err := pachClient.PfsAPIClient.CreateRepo(pachClient.Ctx(), &pfs.CreateRepoRequest{Repo: specRepo})
```